## Mandacaru.Dev 
Atividade Individual 2 - Modulo Cacto - Gustavo Aldama M S P


Os Estados com os quais desenvolvi a atividade foram Minas Gerais e Espirito Santo. 

1) Exiba dois box plots lado a lado, um para cada estado, dos índices solicitados:

a - Rendimento nominal médio mensal per capita (3974, censo Demográfico 2010)


Chamada da API do IBGE, normalização do json e visualização da resposta: 

In [ ]:
import requests
import json
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt

url_1a = 'https://servicodados.ibge.gov.br/api/v3/agregados/3974/periodos/2010/variaveis/3948?localidades=N3[31,32]|N6[N3[31,32]]&classificacao=12085[100543]|58[95253]'

response = requests.get(url_1a)


In [ ]:
resp = response.json()
print(resp)

[{'id': '3948', 'variavel': 'Valor do rendimento nominal médio mensal domiciliar per capita das pessoas de 10 anos ou mais de idade, com rendimento mensal domiciliar per capita, residentes em domicílios particulares, exclusive as pessoas cuja condição no domicílio era pensionista, empregado(a) doméstico(a) ou parente do(a) empregado(a) doméstico(a)', 'unidade': 'Reais', 'resultados': [{'classificacoes': [{'id': '12085', 'nome': 'Situação de ocupação na semana de referência', 'categoria': {'100543': 'Total'}}, {'id': '58', 'nome': 'Grupo de idade', 'categoria': {'95253': 'Total'}}], 'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'2010': '793.2'}}, {'localidade': {'id': '32', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Espírito Santo'}, 'serie': {'2010': '868.7'}}, {'localidade': {'id': '3100104', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abadia dos Dourados - MG'}, 'serie': {

In [ ]:
for item in resp:
  for key in item['resultados']:
    respJson = key
respJson.pop('classificacoes')

[{'categoria': {'100543': 'Total'},
  'id': '12085',
  'nome': 'Situação de ocupação na semana de referência'},
 {'categoria': {'95253': 'Total'}, 'id': '58', 'nome': 'Grupo de idade'}]

In [ ]:
print(respJson)

{'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'2010': '793.2'}}, {'localidade': {'id': '32', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Espírito Santo'}, 'serie': {'2010': '868.7'}}, {'localidade': {'id': '3100104', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abadia dos Dourados - MG'}, 'serie': {'2010': '632.2'}}, {'localidade': {'id': '3100203', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abaeté - MG'}, 'serie': {'2010': '756.7'}}, {'localidade': {'id': '3100302', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abre Campo - MG'}, 'serie': {'2010': '485.0'}}, {'localidade': {'id': '3100401', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Acaiaca - MG'}, 'serie': {'2010': '362.6'}}, {'localidade': {'id': '3100500', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Açucena - MG'}, 'serie': {'2010': '354.9'}}, {'localidade': {'id': '3100609', 'nivel': {

Tratamento dos dados com a biblioteca Pandas para construção dos Boxplots:

In [ ]:
RendaPerCapita_MG_ES = pd.json_normalize(respJson['series'])
RendaPerCapita_MG_ES.head()

localidade.id localidade.nivel.id localidade.nivel.nome  \
0            31                  N3  Unidade da Federação   
1            32                  N3  Unidade da Federação   
2       3100104                  N6             Município   
3       3100203                  N6             Município   
4       3100302                  N6             Município   

            localidade.nome serie.2010  
0              Minas Gerais      793.2  
1            Espírito Santo      868.7  
2  Abadia dos Dourados - MG      632.2  
3               Abaeté - MG      756.7  
4           Abre Campo - MG      485.0

In [ ]:
RendaPerCapita_MG_ES.drop(columns=[RendaPerCapita_MG_ES.columns[1],RendaPerCapita_MG_ES.columns[2]],inplace=True)
RendaPerCapita_MG_ES = RendaPerCapita_MG_ES.drop([0,1], axis='index') 
print(RendaPerCapita_MG_ES)

    localidade.id           localidade.nome serie.2010
2         3100104  Abadia dos Dourados - MG      632.2
3         3100203               Abaeté - MG      756.7
4         3100302           Abre Campo - MG      485.0
5         3100401              Acaiaca - MG      362.6
6         3100500              Açucena - MG      354.9
..            ...                       ...        ...
928       3205101                Viana - ES      544.3
929       3205150           Vila Pavão - ES      478.7
930       3205176         Vila Valério - ES      610.1
931       3205200           Vila Velha - ES     1280.5
932       3205309              Vitória - ES     1981.4

[931 rows x 3 columns]


In [ ]:
RendaPerCapita_MG_ES.rename(columns = {RendaPerCapita_MG_ES.columns[0]:'ID', 
                                     RendaPerCapita_MG_ES.columns[1]:'CIDADE', 
                                     RendaPerCapita_MG_ES.columns[2]: 'RENDA PER CAPITA'}, inplace = True)
RendaPerCapita_MG_ES.set_index('ID', inplace = True)
RendaPerCapita_MG_ES.head()

CIDADE RENDA PER CAPITA
ID                                                
3100104  Abadia dos Dourados - MG            632.2
3100203               Abaeté - MG            756.7
3100302           Abre Campo - MG            485.0
3100401              Acaiaca - MG            362.6
3100500              Açucena - MG            354.9

In [ ]:
RendaPerCapita_MG_ES[['CIDADE','UF']] = RendaPerCapita_MG_ES['CIDADE'].str.split(' - ',1).tolist()
RendaPerCapita_MG_ES.head()

CIDADE RENDA PER CAPITA  UF
ID                                               
3100104  Abadia dos Dourados            632.2  MG
3100203               Abaeté            756.7  MG
3100302           Abre Campo            485.0  MG
3100401              Acaiaca            362.6  MG
3100500              Açucena            354.9  MG

Com os dados tratados com a biblioteca Pandas, gerei duas lista, um para cada estado, para construir os Boxplots:

In [ ]:
RendaPerCapita_MG = []
RendaPerCapita_ES = []
for i in range(len(RendaPerCapita_MG_ES)):
  if str(RendaPerCapita_MG_ES['UF'][i]) == 'MG':
    RendaPerCapita_MG.append(RendaPerCapita_MG_ES['RENDA PER CAPITA'][i])
  else:
    RendaPerCapita_ES.append(RendaPerCapita_MG_ES['RENDA PER CAPITA'][i])

In [ ]:
RendaPerCapita_MG = [float(i) for i in RendaPerCapita_MG]
RendaPerCapita_ES = [float(i) for i in RendaPerCapita_ES]
print(RendaPerCapita_MG)

data_1a = []
data_1a.append(RendaPerCapita_MG)
data_1a.append(RendaPerCapita_ES)
print(data_1a)

[632.2, 756.7, 485.0, 362.6, 354.9, 348.7, 713.7, 560.3, 419.2, 330.3, 589.6, 568.5, 432.8, 593.5, 719.9, 817.2, 416.3, 451.4, 420.4, 663.1, 515.9, 533.7, 349.2, 348.5, 490.3, 315.2, 396.2, 786.8, 288.6, 558.4, 347.6, 485.2, 411.5, 551.4, 515.8, 367.8, 426.4, 841.7, 439.9, 353.7, 559.5, 729.9, 627.4, 922.9, 607.5, 723.9, 663.6, 468.3, 279.1, 391.8, 604.8, 345.3, 406.4, 533.9, 495.6, 813.6, 326.1, 578.0, 576.1, 404.2, 903.6, 390.8, 570.4, 468.5, 493.3, 1561.4, 493.4, 560.3, 321.3, 327.4, 274.3, 699.2, 362.1, 782.1, 591.8, 622.3, 524.3, 487.7, 848.4, 490.7, 623.0, 498.1, 390.7, 460.0, 582.2, 535.7, 497.3, 219.1, 660.1, 662.8, 250.5, 480.4, 378.9, 319.1, 448.8, 634.3, 917.3, 522.9, 491.0, 343.9, 528.7, 365.0, 458.5, 616.4, 600.6, 560.7, 625.6, 555.2, 624.9, 439.8, 354.2, 655.0, 446.2, 686.8, 747.3, 629.8, 377.1, 737.6, 717.8, 704.0, 267.1, 680.7, 496.6, 840.5, 613.2, 543.7, 476.6, 758.7, 491.7, 648.5, 346.1, 391.3, 359.9, 474.9, 588.5, 519.5, 903.9, 427.8, 308.2, 693.2, 389.3, 274.6, 337.

Plotagem dos Boxplots:

In [ ]:
import plotly.graph_objects as go

fig = go.Figure()
names = ['Minas Gerais','Espírito Santo']
for i in range(0,2):
    fig.add_trace(go.Box(y=data_1a[i],name=names[i]))
fig.update_layout(yaxis_title='Renda Per Capita(R$)')
fig.show()

O gráfico acima mostra a semelhança entre a distruição da renda per capita dos dois estados(estatísticas semelhantes), sendo Minas Gerais apresentando mais outliers que Espírito Santo(apenas dois, sendo o maior sem dúvidas a capital do Estado, Vitória).


b - Número de cabeças de bovinos Total (1224 - Censo Agropecuário 2006)
obs: as amostras do box plot devem corresponder às cidades de cada estado.

Chamada da API do IBGE, normalização do json e visualização da resposta: 

In [ ]:
url_1b = 'https://servicodados.ibge.gov.br/api/v3/agregados/1224/periodos/2006/variaveis/2057?localidades=N6[N3[31,32]]&classificacao=12947[0]'

response = requests.get(url_1b)

In [ ]:
resp = response.json()
print(resp)

[{'id': '2057', 'variavel': 'Número de cabeças de bovinos nos estabelecimentos agropecuários', 'unidade': 'Cabeças', 'resultados': [{'classificacoes': [{'id': '12947', 'nome': 'Tipo de agricultura', 'categoria': {'0': 'Total'}}], 'series': [{'localidade': {'id': '3100104', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abadia dos Dourados - MG'}, 'serie': {'2006': '48080'}}, {'localidade': {'id': '3100203', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abaeté - MG'}, 'serie': {'2006': '98504'}}, {'localidade': {'id': '3100302', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abre Campo - MG'}, 'serie': {'2006': '19603'}}, {'localidade': {'id': '3100401', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Acaiaca - MG'}, 'serie': {'2006': '4155'}}, {'localidade': {'id': '3100500', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Açucena - MG'}, 'serie': {'2006': '29337'}}, {'localidade': {'id': '3100609', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Água 

In [ ]:
for item in resp:
  for key in item['resultados']:
    respJson = key
respJson.pop('classificacoes')

[{'categoria': {'0': 'Total'}, 'id': '12947', 'nome': 'Tipo de agricultura'}]

In [ ]:
print(respJson)

{'series': [{'localidade': {'id': '3100104', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abadia dos Dourados - MG'}, 'serie': {'2006': '48080'}}, {'localidade': {'id': '3100203', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abaeté - MG'}, 'serie': {'2006': '98504'}}, {'localidade': {'id': '3100302', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Abre Campo - MG'}, 'serie': {'2006': '19603'}}, {'localidade': {'id': '3100401', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Acaiaca - MG'}, 'serie': {'2006': '4155'}}, {'localidade': {'id': '3100500', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Açucena - MG'}, 'serie': {'2006': '29337'}}, {'localidade': {'id': '3100609', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Água Boa - MG'}, 'serie': {'2006': '54897'}}, {'localidade': {'id': '3100708', 'nivel': {'id': 'N6', 'nome': 'Município'}, 'nome': 'Água Comprida - MG'}, 'serie': {'2006': '12991'}}, {'localidade': {'id': '3100807', 'nivel': {'id': 'N

Tratamento dos dados com a biblioteca Pandas para construção dos Boxplots:

In [ ]:
NumCabecasBoi_MG_ES = pd.json_normalize(respJson['series'])
NumCabecasBoi_MG_ES.head()

localidade.id localidade.nivel.id localidade.nivel.nome  \
0       3100104                  N6             Município   
1       3100203                  N6             Município   
2       3100302                  N6             Município   
3       3100401                  N6             Município   
4       3100500                  N6             Município   

            localidade.nome serie.2006  
0  Abadia dos Dourados - MG      48080  
1               Abaeté - MG      98504  
2           Abre Campo - MG      19603  
3              Acaiaca - MG       4155  
4              Açucena - MG      29337

In [ ]:
NumCabecasBoi_MG_ES.drop(columns=[NumCabecasBoi_MG_ES.columns[1],NumCabecasBoi_MG_ES.columns[2]],inplace=True)
NumCabecasBoi_MG_ES = NumCabecasBoi_MG_ES.drop([0,1], axis='index') 
print(NumCabecasBoi_MG_ES)

    localidade.id     localidade.nome serie.2006
2         3100302     Abre Campo - MG      19603
3         3100401        Acaiaca - MG       4155
4         3100500        Açucena - MG      29337
5         3100609       Água Boa - MG      54897
6         3100708  Água Comprida - MG      12991
..            ...                 ...        ...
925       3205101          Viana - ES      20693
926       3205150     Vila Pavão - ES      30504
927       3205176   Vila Valério - ES       7830
928       3205200     Vila Velha - ES      11988
929       3205309        Vitória - ES        224

[928 rows x 3 columns]


In [ ]:
NumCabecasBoi_MG_ES.rename(columns = {NumCabecasBoi_MG_ES.columns[0]:'ID', 
                                     NumCabecasBoi_MG_ES.columns[1]:'CIDADE', 
                                     NumCabecasBoi_MG_ES.columns[2]: 'NUMERO DE CABECAS BOVINAS'}, inplace = True)
NumCabecasBoi_MG_ES.set_index('ID', inplace = True)
NumCabecasBoi_MG_ES.head()

CIDADE NUMERO DE CABECAS BOVINAS
ID                                                   
3100302     Abre Campo - MG                     19603
3100401        Acaiaca - MG                      4155
3100500        Açucena - MG                     29337
3100609       Água Boa - MG                     54897
3100708  Água Comprida - MG                     12991

In [ ]:
NumCabecasBoi_MG_ES[['CIDADE','UF']] = NumCabecasBoi_MG_ES['CIDADE'].str.split(' - ',1).tolist()
NumCabecasBoi_MG_ES.head()

CIDADE NUMERO DE CABECAS BOVINAS  UF
ID                                                  
3100302     Abre Campo                     19603  MG
3100401        Acaiaca                      4155  MG
3100500        Açucena                     29337  MG
3100609       Água Boa                     54897  MG
3100708  Água Comprida                     12991  MG

In [ ]:
NumCabecasBoi_MG = []
NumCabecasBoi_ES = []
for i in range(len(NumCabecasBoi_MG_ES)):
  if str(NumCabecasBoi_MG_ES['UF'][i]) == 'MG':
    NumCabecasBoi_MG.append(NumCabecasBoi_MG_ES['NUMERO DE CABECAS BOVINAS'][i])
  else:
    NumCabecasBoi_ES.append(NumCabecasBoi_MG_ES['NUMERO DE CABECAS BOVINAS'][i])

Havia uma string "X" dos dados, que são numéricos, e optei por retirar dos dados:

In [ ]:
print(NumCabecasBoi_MG)
NumCabecasBoi_MG.remove('X')

['19603', '4155', '29337', '54897', '12991', '10762', '50130', '15875', '91593', '20052', '6882', '1384', '27337', '39155', '3575', '86392', '16295', '27375', '19698', '1129', '18117', '10810', '16433', '9878', '5805', '19432', '19604', '25138', '2590', '12650', '9807', '6175', '8354', '4423', '29638', '83632', '2026', '6787', '7321', '12012', '18124', '49438', '13384', '24354', '12688', '6925', '5956', '121086', '6380', '106604', '10743', '22494', '17285', '82072', '22553', '3155', '1793', '16803', '17770', '20880', '3179', '1110', '14426', '592', '6915', '12649', '12791', '30330', '6624', '9912', '9974', '8058', '19912', '27676', '8764', '43270', '64438', '6438', '13083', '7576', '17797', '12623', '34331', '8936', '42251', '13660', '16335', '13543', '9097', '49414', '37118', '4710', '13212', '15225', '15331', '18600', '12844', '11526', '102256', '146270', '30794', '21305', '2471', '15344', '10675', '6309', '6944', '3876', '2358', '25115', '6242', '9667', '13275', '11538', '20413', '1

Com os dados tratados com a biblioteca Pandas, gerei duas lista, um para cada estado, para construir os Boxplots:

In [ ]:
NumCabecasBoi_MG = [int(i) for i in NumCabecasBoi_MG]
NumCabecasBoi_ES = [int(i) for i in NumCabecasBoi_ES]
print(NumCabecasBoi_MG)

data_1b = []
data_1b.append(NumCabecasBoi_MG)
data_1b.append(NumCabecasBoi_ES)
print(data_1b)

[19603, 4155, 29337, 54897, 12991, 10762, 50130, 15875, 91593, 20052, 6882, 1384, 27337, 39155, 3575, 86392, 16295, 27375, 19698, 1129, 18117, 10810, 16433, 9878, 5805, 19432, 19604, 25138, 2590, 12650, 9807, 6175, 8354, 4423, 29638, 83632, 2026, 6787, 7321, 12012, 18124, 49438, 13384, 24354, 12688, 6925, 5956, 121086, 6380, 106604, 10743, 22494, 17285, 82072, 22553, 3155, 1793, 16803, 17770, 20880, 3179, 1110, 14426, 592, 6915, 12649, 12791, 30330, 6624, 9912, 9974, 8058, 19912, 27676, 8764, 43270, 64438, 6438, 13083, 7576, 17797, 12623, 34331, 8936, 42251, 13660, 16335, 13543, 9097, 49414, 37118, 4710, 13212, 15225, 15331, 18600, 12844, 11526, 102256, 146270, 30794, 21305, 2471, 15344, 10675, 6309, 6944, 3876, 2358, 25115, 6242, 9667, 13275, 11538, 20413, 14646, 23896, 322387, 15036, 30406, 5022, 39728, 25226, 25678, 4602, 42365, 15430, 27521, 7330, 1802, 3181, 11911, 15078, 7419, 28107, 14303, 54654, 18387, 3822, 17385, 6030, 11637, 14733, 25661, 8306, 9825, 218907, 6753, 18078, 271

Plotagem dos Boxplots:

In [ ]:
import plotly.graph_objects as go

fig = go.Figure()
names = ['Minas Gerais','Espírito Santo']
for i in range(0,2):
    fig.add_trace(go.Box(y=data_1b[i],name=names[i]))
fig.update_layout(yaxis_title='Número de Cabeças de Gado')
fig.show()

O gráfico acima mostra a semelhança entre a distruição da renda per capita dos dois estados(estatísticas semelhantes), explicado pelas semelhanças de clima e vegetação entre eles(elementos essenciais para criação bovina), e como Minas Gerais apresenta maior extensão territorial, aumenta a variabilidade das características naturais, explicando assim seu maior número de outliers.

2) Exiba um gráfico de linha, onde cada linha corresponda a um estado, com a informação de:

a - Índices da Construção Civil: Custo médio (2296) por número índice (49) a partir de 1º de Janeiro de 2000

 

Chamada da API do IBGE, normalização do json e visualização da resposta:

In [ ]:
url_2a = 'https://servicodados.ibge.gov.br/api/v3/agregados/2296/periodos/200001|200002|200003|200004|200005|200006|200007|200008|200009|200010|200011|200012|200101|200102|200103|200104|200105|200106|200107|200108|200109|200110|200111|200112|200201|200202|200203|200204|200205|200206|200207|200208|200209|200210|200211|200212|200301|200302|200303|200304|200305|200306|200307|200308|200309|200310|200311|200312|200401|200402|200403|200404|200405|200406|200407|200408|200409|200410|200411|200412|200501|200502|200503|200504|200505|200506|200507|200508|200509|200510|200511|200512|200601|200602|200603|200604|200605|200606|200607|200608|200609|200610|200611|200612|200701|200702|200703|200704|200705|200706|200707|200708|200709|200710|200711|200712|200801|200802|200803|200804|200805|200806|200807|200808|200809|200810|200811|200812|200901|200902|200903|200904|200905|200906|200907|200908|200909|200910|200911|200912|201001|201002|201003|201004|201005|201006|201007|201008|201009|201010|201011|201012|201101|201102|201103|201104|201105|201106|201107|201108|201109|201110|201111|201112|201201|201202|201203|201204|201205|201206|201207|201208|201209|201210|201211|201212|201301|201302|201303|201304|201305|201306|201307|201308|201309|201310|201311|201312|201401|201402|201403|201404|201405|201406|201407|201408|201409|201410|201411|201412|201501|201502|201503|201504|201505|201506|201507|201508|201509|201510|201511|201512|201601|201602|201603|201604|201605|201606|201607|201608|201609|201610|201611|201612|201701|201702|201703|201704|201705|201706|201707|201708|201709|201710|201711|201712|201801|201802|201803|201804|201805|201806|201807|201808|201809|201810|201811|201812|201901|201902|201903|201904|201905|201906|201907|201908|201909|201910|201911|201912|202001|202002|202003|202004|202005|202006|202007|202008|202009|202010|202011|202012|202101|202102|202103|202104|202105|202106|202107|202108|202109|202110|202111|202112|202201|202202|202203|202204/variaveis/49?localidades=N3[31,32]'
response = requests.get(url_2a)

In [ ]:
resp = response.json()
print(resp)

[{'id': '49', 'variavel': 'Custo médio m² - número-índice', 'unidade': 'Número-índice', 'resultados': [{'classificacoes': [], 'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'200001': '154.02', '200002': '154.84', '200003': '156.03', '200004': '156.67', '200005': '157.25', '200006': '157.42', '200007': '157.72', '200008': '158.40', '200009': '158.43', '200010': '158.84', '200011': '158.92', '200012': '159.33', '200101': '165.02', '200102': '166.38', '200103': '166.56', '200104': '166.74', '200105': '167.37', '200106': '168.83', '200107': '170.13', '200108': '171.76', '200109': '173.36', '200110': '174.21', '200111': '174.73', '200112': '182.21', '200201': '183.94', '200202': '184.94', '200203': '185.14', '200204': '185.58', '200205': '185.86', '200206': '186.29', '200207': '187.24', '200208': '188.18', '200209': '189.59', '200210': '191.22', '200211': '195.12', '200212': '207.88', '200301': '212.70', '2003

In [ ]:
for item in resp:
  for key in item['resultados']:
    respJson = key
respJson.pop('classificacoes')

[]

In [ ]:
print(respJson)

{'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'200001': '154.02', '200002': '154.84', '200003': '156.03', '200004': '156.67', '200005': '157.25', '200006': '157.42', '200007': '157.72', '200008': '158.40', '200009': '158.43', '200010': '158.84', '200011': '158.92', '200012': '159.33', '200101': '165.02', '200102': '166.38', '200103': '166.56', '200104': '166.74', '200105': '167.37', '200106': '168.83', '200107': '170.13', '200108': '171.76', '200109': '173.36', '200110': '174.21', '200111': '174.73', '200112': '182.21', '200201': '183.94', '200202': '184.94', '200203': '185.14', '200204': '185.58', '200205': '185.86', '200206': '186.29', '200207': '187.24', '200208': '188.18', '200209': '189.59', '200210': '191.22', '200211': '195.12', '200212': '207.88', '200301': '212.70', '200302': '215.91', '200303': '216.30', '200304': '217.99', '200305': '218.95', '200306': '220.35', '200307': '221.06', '200308': 

Tratamento dos dados com a biblioteca Pandas para construção dos gráficos de linha:


In [ ]:
CustoMedio_MG_ES = pd.json_normalize(respJson['series'])
CustoMedio_MG_ES.head()

localidade.id localidade.nivel.id localidade.nivel.nome localidade.nome  \
0            31                  N3  Unidade da Federação    Minas Gerais   
1            32                  N3  Unidade da Federação  Espírito Santo   

  serie.200001 serie.200002 serie.200003 serie.200004 serie.200005  \
0       154.02       154.84       156.03       156.67       157.25   
1       141.42       141.97       143.58       143.76       143.92   

  serie.200006  ... serie.202107 serie.202108 serie.202109 serie.202110  \
0       157.42  ...       787.61       795.48       796.76       803.85   
1       144.05  ...       748.64       750.96       753.36       757.95   

  serie.202111 serie.202112 serie.202201 serie.202202 serie.202203  \
0       806.26       806.66       810.54       817.10       848.48   
1       770.54       780.71       788.98       791.11       797.44   

  serie.202204  
0       862.82  
1       807.17  

[2 rows x 272 columns]

In [ ]:
CustoMedio_MG_ES=CustoMedio_MG_ES.drop(["localidade.id","localidade.nivel.id","localidade.nivel.nome","localidade.nome"], axis ='columns')

Com os dados tratados com a biblioteca Pandas, gerei duas lista, um para cada estado, para construir os gráficos de linha:

In [ ]:
lista = CustoMedio_MG_ES.values
CustoMedio_MG = [float(x) for x in lista[0]]
CustoMedio_ES = [float(x) for x in lista[1]]

Tratamento das datas para índice no eixo x do gráfico:

In [ ]:
indice = list(CustoMedio_MG_ES.columns)
for i in range(len(indice)):
   indice[i] = indice[i][6:10]+'-'+indice[i][10:]

print(indice)

['2000-01', '2000-02', '2000-03', '2000-04', '2000-05', '2000-06', '2000-07', '2000-08', '2000-09', '2000-10', '2000-11', '2000-12', '2001-01', '2001-02', '2001-03', '2001-04', '2001-05', '2001-06', '2001-07', '2001-08', '2001-09', '2001-10', '2001-11', '2001-12', '2002-01', '2002-02', '2002-03', '2002-04', '2002-05', '2002-06', '2002-07', '2002-08', '2002-09', '2002-10', '2002-11', '2002-12', '2003-01', '2003-02', '2003-03', '2003-04', '2003-05', '2003-06', '2003-07', '2003-08', '2003-09', '2003-10', '2003-11', '2003-12', '2004-01', '2004-02', '2004-03', '2004-04', '2004-05', '2004-06', '2004-07', '2004-08', '2004-09', '2004-10', '2004-11', '2004-12', '2005-01', '2005-02', '2005-03', '2005-04', '2005-05', '2005-06', '2005-07', '2005-08', '2005-09', '2005-10', '2005-11', '2005-12', '2006-01', '2006-02', '2006-03', '2006-04', '2006-05', '2006-06', '2006-07', '2006-08', '2006-09', '2006-10', '2006-11', '2006-12', '2007-01', '2007-02', '2007-03', '2007-04', '2007-05', '2007-06', '2007-07'

Plotagem dos gráficos de linha:

In [ ]:
figPx = go.Figure()
figPx.add_trace(go.Scatter(x=indice, y=CustoMedio_MG, name='Minas Gerais',
                         line=dict(color='blue', width=1)))
figPx.add_trace(go.Scatter(x=indice, y=CustoMedio_ES, name='Espirito Santo',
                         line=dict(color='firebrick', width=1)))

figPx.update_traces(hovertemplate='Value: %{y:.2f}')
figPx.update_layout(hovermode="x unified",
                        plot_bgcolor='rgba(0,0,0,0)',
                        title={'x':0.5,'xanchor': 'center',
                              'font':{'size': 24}
                        },
                        xaxis=dict(
                            title="Ano",
                            showgrid=False,  # Removes X-axis grid lines
                            dtick= 'M48'
                        ),
                        yaxis=dict(
                            title="Custo Médio por Metro Quadrado",  
                            tickformat = ',f',  # Add thousands
                            showgrid=False,  # Removes Y-axis grid lines    
                            zeroline=False
                        ))


figPx.add_annotation(
            x=dt.datetime(2020, 2, 1),
            y=550,
            xref="x",
            yref="y",
            text="<b>Declaração da Pandemia</b><br><b> da Covid-19</b><br><b> (Março 2020)</b>",
            showarrow=True,
            arrowhead=5,
            ax=80,
            ay=60)

No gráfico vemos um aumentos quase linar no preço do metro quadrado na contrução civil, fato decorrente da inflação anual e mercado aquecido. Por outro lado, com a declaração da pandemia da covid-19, vemos uma retração nesse setor, e uma escassez de materiais devido a quebra na cadeia de suprimentos, o que causou um crescimentos também linear, mas a uma taxa maior(maior inclinação do gráfico de linha).


b - Produto Interno Bruto: Índice Gini do PIB, referência 2010 (5939, 529) para cada estado, a partir de 2002

Chamada da API do IBGE, normalização do json e visualização da resposta:

In [ ]:
url_2b = 'https://servicodados.ibge.gov.br/api/v3/agregados/5939/periodos/2002|2003|2004|2005|2006|2007|2008|2009|2010|2011|2012|2013|2014|2015|2016|2017|2018|2019/variaveis/529?localidades=N3[31,32]'
response = requests.get(url_2b)

In [ ]:
resp = response.json()
print(resp)

[{'id': '529', 'variavel': 'Índice de Gini da distribuição do produto interno bruto a preços correntes', 'unidade': 'Índice', 'resultados': [{'classificacoes': [], 'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'2002': '0.822643', '2003': '0.823687', '2004': '0.826222', '2005': '0.824006', '2006': '0.825257', '2007': '0.831472', '2008': '0.831392', '2009': '0.829374', '2010': '0.833219', '2011': '0.827357', '2012': '0.826352', '2013': '0.829611', '2014': '0.825613', '2015': '0.819167', '2016': '0.810631', '2017': '0.814644', '2018': '0.818877', '2019': '0.818079'}}, {'localidade': {'id': '32', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Espírito Santo'}, 'serie': {'2002': '0.787016', '2003': '0.787974', '2004': '0.798309', '2005': '0.795508', '2006': '0.792566', '2007': '0.793213', '2008': '0.799044', '2009': '0.787945', '2010': '0.787483', '2011': '0.783108', '2012': '0.774786', '2013

In [ ]:
for item in resp:
  for key in item['resultados']:
    respJson = key
respJson.pop('classificacoes')

[]

In [ ]:
print(respJson)

{'series': [{'localidade': {'id': '31', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Minas Gerais'}, 'serie': {'2002': '0.822643', '2003': '0.823687', '2004': '0.826222', '2005': '0.824006', '2006': '0.825257', '2007': '0.831472', '2008': '0.831392', '2009': '0.829374', '2010': '0.833219', '2011': '0.827357', '2012': '0.826352', '2013': '0.829611', '2014': '0.825613', '2015': '0.819167', '2016': '0.810631', '2017': '0.814644', '2018': '0.818877', '2019': '0.818079'}}, {'localidade': {'id': '32', 'nivel': {'id': 'N3', 'nome': 'Unidade da Federação'}, 'nome': 'Espírito Santo'}, 'serie': {'2002': '0.787016', '2003': '0.787974', '2004': '0.798309', '2005': '0.795508', '2006': '0.792566', '2007': '0.793213', '2008': '0.799044', '2009': '0.787945', '2010': '0.787483', '2011': '0.783108', '2012': '0.774786', '2013': '0.768602', '2014': '0.769294', '2015': '0.757844', '2016': '0.747877', '2017': '0.744252', '2018': '0.769704', '2019': '0.762399'}}]}


Tratamento dos dados com a biblioteca Pandas para construção dos gráficos de linha:

In [ ]:
PibGini_MG_ES = pd.json_normalize(respJson['series'])
PibGini_MG_ES.head()

localidade.id localidade.nivel.id localidade.nivel.nome localidade.nome  \
0            31                  N3  Unidade da Federação    Minas Gerais   
1            32                  N3  Unidade da Federação  Espírito Santo   

  serie.2002 serie.2003 serie.2004 serie.2005 serie.2006 serie.2007  ...  \
0   0.822643   0.823687   0.826222   0.824006   0.825257   0.831472  ...   
1   0.787016   0.787974   0.798309   0.795508   0.792566   0.793213  ...   

  serie.2010 serie.2011 serie.2012 serie.2013 serie.2014 serie.2015  \
0   0.833219   0.827357   0.826352   0.829611   0.825613   0.819167   
1   0.787483   0.783108   0.774786   0.768602   0.769294   0.757844   

  serie.2016 serie.2017 serie.2018 serie.2019  
0   0.810631   0.814644   0.818877   0.818079  
1   0.747877   0.744252   0.769704   0.762399  

[2 rows x 22 columns]

In [ ]:
PibGini_MG_ES=PibGini_MG_ES.drop(["localidade.id","localidade.nivel.id","localidade.nivel.nome","localidade.nome"], axis ='columns')

Com os dados tratados com a biblioteca Pandas, gerei duas lista, um para cada estado, para construir os gráficos de linha:

In [ ]:
lista2 = PibGini_MG_ES.values
PibGini_MG = [float(x) for x in lista2[0]]
PibGini_ES = [float(x) for x in lista2[1]]

Tratamento das datas para índice no eixo x do gráfico:

In [ ]:
indice2 = list(PibGini_MG_ES.columns)
for i in range(len(indice2)):
   indice2[i] = indice2[i][6:10]

print(indice2)

['2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019']


Plotagem dos gráficos de linha:

In [ ]:
figPx = go.Figure()
figPx.add_trace(go.Scatter(x=indice2, y=PibGini_MG, name='Minas Gerais',
                         line=dict(color='blue', width=1)))
figPx.add_trace(go.Scatter(x=indice2, y=PibGini_ES, name='Espirito Santo',
                         line=dict(color='firebrick', width=1)))

figPx.update_traces(hovertemplate='Value: %{y:.2f}')
figPx.update_layout(hovermode="x unified",
                        plot_bgcolor='rgba(0,0,0,0)',
                        title={'x':0.5,'xanchor': 'center',
                              'font':{'size': 24}
                        },
                        xaxis=dict(
                            title="Ano",
                            showgrid=False,  # Removes X-axis grid lines
                            dtick= 'M48'
                        ),
                        yaxis=dict(
                            title="Índice Gini do PIB",  
                            tickformat = ',f',  # Add thousands
                            showgrid=False,  # Removes Y-axis grid lines    
                            zeroline=False
                        ))

No gráfico podemos notar que o Índice Gini do PIB de Minas gerais Supera com folga o de Espírito Santo, fato que corrobora o processo histórico-econômico do nosso país, onde por muito tempo Minas Gerais foi o centro econômico do País devido a exploração de metais preciosos, e depois de grande latifúndios, perpertuando assim a concetração de riqueza no estado.

3) Crie uma visualização em mapa coroplético do número de fundações privadas e associações sem fins lucrativos:

a - Número total de unidades (6917 - 2122) por município, no ano de 2016

Chamada da API do IBGE para as malhas dos estados, normalização dos geojson e visualização da resposta:




In [ ]:
#Obtendo os shapes
url_MG = 'https://servicodados.ibge.gov.br/api/v3/malhas/estados/31?formato=application/vnd.geo+json&qualidade=minima&intrarregiao=municipio'
response = requests.get(url_MG)
shapesJson_MG = response.json()


#Obtendo os shapes
url_ES = 'https://servicodados.ibge.gov.br/api/v3/malhas/estados/32?formato=application/vnd.geo+json&qualidade=minima&intrarregiao=municipio'
response = requests.get(url_ES)
shapesJson_ES = response.json()

Junção dos shapes do dois estados:

In [ ]:
shapesJson = shapesJson_MG
shapesJson['features'] = shapesJson_MG['features'] + shapesJson_ES['features']

Chamada da API do IBGE, normalização do json e visualização da resposta e tratamento com Pandas:

In [ ]:
url_3a = 'https://servicodados.ibge.gov.br/api/v3/agregados/6917/periodos/2016/variaveis/2122?localidades=N6[N3[31,32]]&classificacao=12665[all]'
response = requests.get(url_3a)
ongs = response.json()
for item in ongs:
  for key in item['resultados']:
    ongsJson = key
ongsJson.pop('classificacoes')
ONGS = pd.json_normalize(ongsJson['series'])
ONGS[['CIDADE','UF']] = ONGS[ONGS.columns[3]].str.split(' - ',1).tolist()
ONGS.drop(columns=[ONGS.columns[1],
                    ONGS.columns[2],
                    ONGS.columns[3]],
           inplace=True)
ONGS.rename(columns = {ONGS.columns[0]:'ID', 
                       ONGS.columns[1]: 'ONGS'},
             inplace = True)
ONGS.set_index('ID', inplace = True)
new_columns = (ONGS.columns.drop('ONGS').tolist()) + ['ONGS']
ONGS = ONGS[new_columns]
ONGS = ONGS.replace('-',0)
ONGS.head()

CIDADE  UF ONGS
ID                                   
3100104  Abadia dos Dourados  MG    1
3100203               Abaeté  MG    0
3100302           Abre Campo  MG    1
3100401              Acaiaca  MG    0
3100500              Açucena  MG    0

In [ ]:
print(ONGS.head)

<bound method NDFrame.head of                       CIDADE  UF ONGS
ID                                   
3100104  Abadia dos Dourados  MG    1
3100203               Abaeté  MG    0
3100302           Abre Campo  MG    1
3100401              Acaiaca  MG    0
3100500              Açucena  MG    0
...                      ...  ..  ...
3205101                Viana  ES    1
3205150           Vila Pavão  ES    2
3205176         Vila Valério  ES    0
3205200           Vila Velha  ES   34
3205309              Vitória  ES  282

[931 rows x 3 columns]>


Havia uma string "..." dos dados, que são numéricos, e optei por substitui-la por 0:


In [ ]:
X = ONGS['ONGS'].replace('...',0)
X.to_numpy
ONGS['ONGS'] = X.astype(float)

Plotagem do mapa cloroplético:

In [ ]:
import plotly.express as px
from plotly.offline import init_notebook_mode, plot, iplot, download_plotlyjs
br_lat = -19
br_lon = -41.0
brazilMap = px.choropleth_mapbox(ONGS,
                geojson=shapesJson, 
                locations=ONGS.index,
                hover_name = 'CIDADE',
                color='ONGS', 
                color_continuous_scale=[(0, "blue"), (0.2, "green"), (1, "red")],
                featureidkey="properties.codarea",
                )
brazilMap.update_layout(mapbox_style="open-street-map", 
                            mapbox_zoom=4.8, 
                            mapbox_center = {"lat": br_lat, "lon": br_lon},
                            title="QUANTIDADE DE INSTITUIÇÕES PRIVADAS E ONGS - MINAS GERAIS E ESPIRITO SANTO")
brazilMap.show()

O gráfico mostra uma distruição das ONGS em um range pequeno, de até 80 (Contagem (MG)) instituições. Por outro lado, temos outliers discrepantes como Belo Horizonte(MG): 471, Juiz de Fora(MG): 578 e Vitória(ES): 282

b - Exiba, também, o número de unidades per capita, utilizando os dados de estimativa populacional em 2021 

Chamada da API do IBGE, normalização do json e visualização da resposta e tratamento com Pandas:

In [ ]:
# Obtendo a densidade populacional: Ceará e Pernambuco
url_3b = "https://servicodados.ibge.gov.br/api/v3/agregados/6579/periodos/2021/variaveis/9324?localidades=N6[N3[31,32]]"

response = requests.get(url_3b)
dens = response.json()
for item in dens:
  for key in item['resultados']:
    densJson = key
densJson.pop('classificacoes')
dfDens = pd.json_normalize(densJson['series'])
dfDens.drop(columns=[dfDens.columns[1],
                    dfDens.columns[2],
                    dfDens.columns[3]],
           inplace=True)
dfDens.rename(columns = {dfDens.columns[0]:'ID', 
                       dfDens.columns[1]: 'POPULACAO'},
             inplace = True)
dfDens.set_index('ID', inplace = True)
dfDens.head()

POPULACAO
ID               
3100104      7022
3100203     23263
3100302     13434
3100401      3995
3100500      9270

In [ ]:
dfONGSDens = pd.merge(ONGS, dfDens, on="ID")
dfONGSDens[['ONGS','POPULACAO']]=dfONGSDens[['ONGS','POPULACAO']].astype(float)
dfONGSDens.head()

CIDADE  UF  ONGS  POPULACAO
ID                                               
3100104  Abadia dos Dourados  MG   1.0     7022.0
3100203               Abaeté  MG   0.0    23263.0
3100302           Abre Campo  MG   1.0    13434.0
3100401              Acaiaca  MG   0.0     3995.0
3100500              Açucena  MG   0.0     9270.0

Gerando Dataframe com a quantidade de ONGS Per Capita:

In [ ]:
dfONGSDens['ONGS']=dfONGSDens['ONGS']/dfONGSDens['POPULACAO']

dfONGSDens.drop(columns=[dfONGSDens.columns[3]],inplace=True)

print(dfONGSDens)

                      CIDADE  UF      ONGS
ID                                        
3100104  Abadia dos Dourados  MG  0.000142
3100203               Abaeté  MG  0.000000
3100302           Abre Campo  MG  0.000074
3100401              Acaiaca  MG  0.000000
3100500              Açucena  MG  0.000000
...                      ...  ..       ...
3205101                Viana  ES  0.000012
3205150           Vila Pavão  ES  0.000216
3205176         Vila Valério  ES  0.000000
3205200           Vila Velha  ES  0.000067
3205309              Vitória  ES  0.000763

[931 rows x 3 columns]


Plotagem do mapa cloroplético:

In [ ]:
import plotly.express as px
from plotly.offline import init_notebook_mode, plot, iplot, download_plotlyjs
br_lat = -19
br_lon = -41.0
brazilMap = px.choropleth_mapbox(dfONGSDens,
                geojson=shapesJson, 
                locations=ONGS.index,
                hover_name = 'CIDADE',
                color='ONGS', 
                color_continuous_scale=[(0, "blue"), (0.2, "green"), (1, "red")],
                featureidkey="properties.codarea",
                )
brazilMap.update_layout(mapbox_style="open-street-map", 
                            mapbox_zoom=4.8, 
                            mapbox_center = {"lat": br_lat, "lon": br_lon},
                            title="QUANTIDADE DE INSTITUIÇÕES PRIVADAS E ONGS PER CAPITA - MINAS GERAIS E ESPÍRITO SANTO")
brazilMap.show()